include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -388,21 +488,275 @@ __single_pass_scan(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __out_r
+        if (oneapi::dpl::__par_backend_hetero::__group_copy_if_fits_in_slm(__queue, __n, __n_uniform))
+        {
+            return oneapi::dpl::__par_backend_hetero::__dispatch_small_copy_if(
+                oneapi::dpl::execution::__dpl::make_device_policy<_KernelName>(__queue), __n,
+                std::forward<_InRng>(__in_rng), std::forward<_OutRng>(__out_rng),
+                std::forward<_NumCopiedRng>(__num_rng), __pred);
+        }
+    }
+    constexpr std::size_t __workgroup_size = _KernelParam::workgroup_size;
+    constexpr std::size_t __elems_per_workitem = _KernelParam::data_per_workitem;
+
+    // Avoid non_uniform n by padding up to a multiple of __workgroup_size
+    constexpr std::uint32_t __elems_in_tile = __workgroup_size * __elems_per_workitem;
+    std::size_t __current_num_wgs = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __elems_in_tile);
+    std::size_t __current_num_items = __current_num_wgs * __workgroup_size;
+
+    __scan_lookback_mem_mgr<_FlagType> __device_mem_mgr(__queue, __current_num_wgs);
+    __device_mem_mgr.allocate();
+
+    // Memory Structure:
+    // [Lookback Scan Memory, Tile Id Counter]
+    auto __status_vals_full = __device_mem_mgr.get_full_values_begin();
+    auto __status_vals_partial = __device_mem_mgr.get_partial_values_begin();
+    auto __status_flags = __device_mem_mgr.get_flags_begin();
+    //adding 1 to the number elements to account for the tile id
+    std::size_t __status_flags_size = __device_mem_mgr.get_num_elements() + 1;
+
+    auto __fill_event = __lookback_init_submitter<_FlagType, _SizeT, _BinaryOp, _LookbackInitKernel>{}(
+        __queue, __status_flags, __status_vals_partial, __status_flags_size, _FlagType::__padding);
+
+    sycl::event __prev_event = __copy_if_submitter<__elems_per_workitem, __workgroup_size, _FlagType, _CopyIfKernel>{}(
+        __queue, __fill_event, __in_rng, __out_rng, __num_rng, __n, __pred, __status_flags, __status_flags_size,
+        __status_vals_full, __status_vals_partial, __current_num_items, __current_num_wgs);
+
+    // TODO: Currently, the following portion of code makes this entire function synchronous.
+    // Ideally, we should be able to use the asynchronous free below, but we have found that doing
+    // so introduces a large unexplainable slowdown. Once this slowdown has been identified and corrected,
+    // we should replace this code with the asynchronous version below.
+    if (0)
+    {
+        return __device_mem_mgr.async_free(__prev_event);
+    }
     else
     {
         __prev_event.wait();
-        sycl::free(__device_mem, __queue);
+        __device_mem_mgr.free();
         return __prev_event;
     }
 }
 
 } // namespace __impl
 
+template <typename _InRng, typename _OutRng, typename _NumCopiedRng, typename _UnaryPredicate, typename _KernelParam>
+sycl::event
+copy_if(sycl::queue __queue, _InRng&& __in_rng, _OutRng&& __out_rng, _NumCopiedRng&& __num_rng, _UnaryPredicate __pred,
```

Comment:
It is natural to see `__num_rng` passed as a pointer rather than a range, because it is expected to hold one element. Will that API be able to digest `__num_rng` as a pointer?

This is definitely not a blocker, because it can be work-arounded on a caller side (e.g. wrapping it with `views::subrange`).

Reply:
My intention here is to allow a buffer of a single element to be easily used here. However, I agree that an iterator / pointer may suit it better, given that it should be a single element.  I don't think this current API works with just a pointer as it is written.

I don't think we should force it to be a pointer, because I don't think we make using buffers difficult, but I think an iterator could be fine (which includes USM pointers, of course).  It would just require users to call `oneapi::dpl::begin(num_copied_buf)` first if they want to use a buffer.

I think thre may also be utility in creating a `single_device_element` type which could wrap a buffer, iterator, USM pointer to represent an single element which can be made available on the device.  Its something I've considered also for the `init` value of scan, where currently we only allow a host-side ValueType value.  If we also had an specialized overload for a `single_device_element` which would allow users to skip the host to device copy. The other place such a type is interesting is for the return of `reduce` for some KT version.

Reply:
Having taken a fresher look, I think that the existing approach below is fine:

```c++
sycl::buffer<T> buf(input.data(), input.size());
sycl::buffer<T> buf_out(size);
sycl::buffer<std::size_t> buf_num_copied(&num_copied, 1);
oneapi::dpl::experimental::kt::gpu::copy_if(q, buf, buf_out, buf_num_copied, pred, param).wait();
```

I previously thought about this one as an alternative:

```c++
sycl::buffer<T> buf(input.data(), input.size());
sycl::buffer<T> buf_out(size);
auto buf_num_copied = sycl::malloc_device<T>(1, q);
oneapi::dpl::experimental::kt::gpu::copy_if(q, buf, buf_out, buf_num_copied, pred, param).wait();
```

Now, it does not look like a good idea to me. 
Users are more likely to stick to either `buffer` or `malloc_device` for all the arguments. I know examples when an external pool of memory of memory is used for all the data, and only pointers are passed. So, it does not justify the additional complexity. Moreover, there are workarounds.

Discussing API of a new experimental feature involves much speculation, sometimes with no strong ground like this one. I hope it is useful at least.

Reply:
> I think thre may also be utility in creating a single_device_element type...

Yes, it will be useful if it simplifies the usage of the features and helps to avoid performance-punishing pitfalls. It would be good to see these cases, though.
What I can add is that introducing extra logic for passing `init` values may not be beneficial, because the latency of `init` value copying may be hidden by the kernel launch. It is a kernel argument, in contrast to return values of `reduce` or `scan` KT algorithms.

Reply:
> What I can add is that introducing extra logic for passing init values may not be beneficial

Actually, it may allow passing non-trivially-copyable type of `init` value in case the algorithm handles data, which is constructed on a device. I do not know such use-cases, though.